Performance optimize include behavior

nbibler opened the pull request:
## What is the current behavior?

Currently (v2.0.0), the logic for loading and serializing included records is highly-inefficient. Further, it appears to have an errant (albeit ignored) behavior. This inefficiency is exacerbated with a large and/or deeply nested set of included objects, which causes exponential growth in object allocation, iteration, and object checking.

Regarding the errant behavior, the current implementation attempts to query a parent object for all nested, included associations. Hard to describe, but using this library's spec fixtures as an example: If I serialize a Movie and ask to `include: [:actors, :'actors.played_movies']`, then the Movie instance is checked for both "actors" and "played_movies". Played movies should not be checked on the movie, only on the associated actors. The current library hides this error with a `next` check, but it still loops and runs. This error is where the exponential iteration growth comes from with deeply nested includes.

Anyway, with the current v2.0.0 release, using a large, deeply nested object set, I was finding results similar to the following when using already-loaded-in-memory object sets (these numbers are strictly for performing the `serializable_hash` call and Oj JSON dump, skipping the database load and whatever else occurred before it):

Serialization (wall) Time: 6090.6ms
Allocations: 14974152

## What is the new behavior?

There is no difference in external behavior. There are, however, significant improvements in object allocation and execution times. After these changes, the same data set and request see numbers like:

Serialization (wall) Time: 769.6ms
Allocations: 1171848

## Checklist

Please make sure the following requirements are complete:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes /
  features)
- [x] All automated checks pass (CI/CD)

stas commented:
@nbibler thank you for the PR!!! Since this is more of a performance related refactoring, may I ask you to share with us an example we could reproduce.

Consider sending a PR (not for merge) at https://github.com/jsonapi-serializer/comparisons
Or just use it to demonstrate the issue.

It should be easy enough to set it up, but it will help us all understand where you're coming from.

nbibler commented:
Sure. I'll take a look. I'm also fixing the failing Rubocop tests.

nbibler commented:
Looking into a possible issue that I may have introduced with the `ObjectSerializer.validate_includes!` method…

nbibler commented:
Alright, I've [forked the comparison repo](https://github.com/nbibler/comparisons/tree/deep-includes) and added an (arguably silly, but demonstrative) deep inclusion test.

Using that, I ran two sets of tests locally. First, it used the comparison repo's Gemfile.lock (which looks to be pointing to 990c0acd18137c96d0c3e4a3168cc71fd3b912c2). Second, I updated the Gemfile to pull from [this PR's ref](https://github.com/jsonapi-serializer/jsonapi-serializer/pull/103/commits/cacca6db1ec142574a2a4ca2489dc4ebb5b087aa)."

What you'll see different between the two runs is that `fast_jsonapi (deep)` goes from being _horribly_ slow (1.5i/s, second slowest of all sets) to moving up to the fasted "deep" test at 21i/s, and expectedly-similar to the `fast_jsonapi (many)` results.

[master.txt](https://github.com/jsonapi-serializer/jsonapi-serializer/files/4941126/master.txt)
[optimizing-includes.txt](https://github.com/jsonapi-serializer/jsonapi-serializer/files/4941127/optimizing-includes.txt)

nbibler commented:
Pinging to see if there's anything left for me to do here.

If there are any concerns about the code or approach, just let me know!

stas commented:
@nbibler apologies for a bit of delay. Let's start by merging your changes to comparisons first:
https://github.com/nbibler/comparisons/commit/95683124a2e88de6838d7bf4faa078571361b257

I'm looking into the PR in the meantime.

stas commented on lib/fast_jsonapi/serialization_core.rb:
```diff
@@ -94,58 +94,47 @@ def id_from_record(record, params)
-        [items[1..-1].join('.').to_sym]
+      def parse_includes_list(includes_list)
+        includes_list.each_with_object({}) do |include_item, include_sets|
+          include_base, include_remainder = include_item.to_s.split('.', 2)
```
Why do we take just the first two split values here?

nbibler replied:
The idea here is that if we're given a long/deep list of includes, like

```ruby
  BookSerializer.new(
    DATA.sample,
    include: [
      'authors',
      'authors.books',
      'authors.books.genre',
      'authors.books.genre.books',
      'authors.books.genre.books.authors',
      'authors.books.genre.books.genre',
      'genre',
      'genre.books'
    ]
  ).serializable_hash
```

The current serializer (BookSerializer, in this case), only cares about the unique, root inclusions requested. In the example above, that would only be `authors`. Everything after authors is outside of the BookSerializer's scope.

This is the bug in the current implementation. Partly, because it chops the include up into a flat array of all nested includes, it asks the BookSerializer to serialize "authors", then "books", then "genre", then "books", then "authors", then "genre", then "genre", then "books." The serialization attempts for BookSerializer relationships exponentially grows with the depth of those includes.

So, back to this implementation: it chops out the root association (first part) from each entry. It uniques it (because of how Hash keys work) and collects all of the "include_remainder" values to hand off to the next appropriate serializer. This method will turn that `include` array into a Hash that looks like:

```ruby
{
  'authors' => Set.new([
    'books',
    'books.genre',
    'books.genre.books',
    'books.genre.books.authors',
    'books.genre.books.genre'
  ]),
  'genre' => Set.new(['books'])
}
```

Because the BookSerializer only cares about the root associations included, it only needs the first segment of each include ("authors" and "genre") and it doesn't need to waste cycles parsing the rest of it. That will be the job of the next serializer in line. For example, the GenreSerializer would only see that it needs to include `["books"]`, while the AuthorsSerializer will get the Set of 5 entries.

Further, this implementation uses `Set`. This gives the benefit of ensuring the list of associations to include are unique. So, if this list is long and the caller mistakenly asked for "books.genre" twice, that relationship will only be iterated (to be checked and serialized) once, because pushing a duplicate entry into a Set is ignored.

stas replied:
Alright, this makes sense. It wasn't clear since we do have further calls to `parse_includes_list`. Do you mind adding a little docstring maybe for the method? If not I can just merge it and update it...

Thanks! :bow:

stas replied:
Added it here 1ce4677

Thanks again @nbibler for all the work here! :bow:

nbibler replied:
Thanks for doing that! I've been swamped this week.

GuiTeK commented:
@nbibler Thank you so much for your PR!

We're using fast-jsonapi (looks like it's been renamed to jsonapi-serializer recently?) in production and we're encountering big performance issues (lots of memory allocations triggering the garbage collector, making everything super slow). We found a few workarounds but it's not practical at all and doesn't fix the problem entirely, so it's still very much a problem for us.

I'm really looking forward to seeing this merged!

nbibler commented:
@stas: I've created a new PR at jsonapi-serializer/comparisons#8.

nbibler commented:
@jopotts Thank you! Happy to help.

stas commented:
> We're using fast-jsonapi (looks like it's been renamed to jsonapi-serializer recently?) in production and we're encountering big performance issues (lots of memory allocations triggering the garbage collector, making everything super slow). We found a few workarounds but it's not practical at all and doesn't fix the problem entirely, so it's still very much a problem for us.

@GuiTeK would it be possible to get a bit more context on how your setup looks like and start a conversation in another issue?! Thanks for sharing the concerns.

matheussilvasantos commented:
I'm afraid `include` lists like `trees.branches.leaves` don't work anymore.
